lib/system/exceptions.nim

```diff
@@ -16,6 +16,10 @@ type
     procname*: cstring      ## Name of the proc that is currently executing.
     line*: int              ## Line number of the proc that is currently executing.
     filename*: cstring      ## Filename of the proc that is currently executing.
+    frameMsg*: string       ## When a stacktrace is generated in a given frame and
+                            ## rendered at a later time, we should ensure the stacktrace
+                            ## data isn't invalidated; any pointer into PFrame is
+                            ## subject to being invalidated so shouldn't be stored.
```

Comment:
Can this even work with the new runtime, it means StackTraceEntry will get an automatic destructor. Also, I'm worried about the overhead. The stack tracing is already so slow that people replace it by "native" solutions that don't have as much overhead.

Reply:
> Can this even work with the new runtime

--gc:arc doesn't handle exceptions correctly (cf issues still opened in nim issues), so `nim c --gc:arc ./tests/stdlib/tstackframes.nim` fails, but for reasons unrelated to this PR; if you remove the `setFrameMsg` calls, it still fails instead of catching properly.
That being said, **it does show the annotated stacktraces** (ie with `setFrameMsg` kept uncommented) even with --gc:arc.

> Also, I'm worried about the overhead.

you only pay for what you use (besides `frameMsgLen`, see below point), that is,  if you use `setFrameMsg` AND if an exception is raised AND if the `setFrameMsg` messages occur between the catcher and the frame that throws. In particular, **unlike initFrame and popFrame for which you pay a price upon entering/exiting each function, here you only pay a price upon throwing**, and when above conditions are met, ie overhead is minimal even when you do use the feature.

```nim
proc codeUsingFeature(i: int)=
  setFrameMsg $i
  doAssert i < 7 # will crash

for i in 0..<100:
  # not affected by this PR so long this doesn't throw, no matter how nested
  # because StackTraceEntry is only used when an exception is thrown, to "serialize" the `PFrame` into something
  # that won't have its pointers invalidated
  hotloop()

for i in 0..<10:
  codeUsingFeature()
```

## note: `frameMsgLen`,
the only unconditional price you pay (with --stacktrace:on, and per frame) is an extra call `s.frameMsgLen = framePtr.frameMsgLen` inside `nimFrame`. If doubt it makes a practical difference, but in subsequent PR, if this gives any kind of meaningful slowdown, we can always add a --stacktraceAnnotation:on/off

## note
if user is careful to serialize the `PFrame` immediately after an exception is caught, the pointers are still valid (because we're in a parent frame from thrown exception) so the `frameMsg` writing can be omitted there, but that's obviously unsafe and making it safe would require future work

Reply:
> --gc:arc doesn't handle exceptions correctly (cf issues still opened in nim issues)

I'm not aware of any open bugs for exception handling with gc:arc.

Reply:
it's hitting that bug: https://github.com/nim-lang/Nim/issues/13585

Reply:
I've modified the test to sidestep  #13585 by intstead throwing `CatchableError`, the test works with or without --gc:arc

Reply:
Why is frameMsgLen needed though? Can't we just use frameMsg.len?